src/thunderbots/software/util/gradient_descent.tpp

```diff
@@ -0,0 +1,150 @@
+{
+    return followGradient(
+        objective_function, initial_value, num_iters,
+        [](double curr_value, double step) { return curr_value + step; });
+}
+
+template <size_t NUM_PARAMS>
+std::array<double, NUM_PARAMS> Util::GradientDescentOptimizer<NUM_PARAMS>::minimize(
+    std::function<double(std::array<double, NUM_PARAMS>)> objective_function,
+    std::array<double, NUM_PARAMS> initial_value, unsigned int num_iters)
+{
+    return followGradient(
+        objective_function, initial_value, num_iters,
+        [](double curr_value, double step) { return curr_value - step; });
+}
+
+template <size_t NUM_PARAMS>
+std::array<double, NUM_PARAMS> Util::GradientDescentOptimizer<NUM_PARAMS>::followGradient(
+    std::function<double(std::array<double, NUM_PARAMS>)> objective_function,
+    std::array<double, NUM_PARAMS> initial_value, unsigned int num_iters,
+    std::function<double(double, double)> gradient_movement_func)
+{
+    // Implementation of the "Adam" algorithm. See Javadoc class comment for this
+    // class (in the header) for details
+
+    // This is basically just to change the name so the below code reads more nicely
+    ParamArray params = initial_value;
+
+    // The past gradient and squared gradient averages for each parameter
+    ParamArray past_gradient_averages         = {0};
+    ParamArray past_squared_gradient_averages = {0};
+
+    for (int iter = 0; iter < num_iters; iter++)
+    {
+        ParamArray gradient = approximateGradient(params, objective_function);
+
+        // Get the squared gradient
+        ParamArray squared_gradient = {0};
+        for (unsigned int i = 0; i < NUM_PARAMS; i++)
+        {
+            squared_gradient.at(i) = std::pow(gradient.at(i), 2);
+        }
+
+        // Update past gradient and gradient squared averages
+        for (unsigned int i = 0; i < NUM_PARAMS; i++)
+        {
+            past_gradient_averages.at(i) =
+                past_gradient_decay_rate * past_gradient_averages.at(i) +
+                (1 - past_gradient_decay_rate) * gradient.at(i);
+            past_squared_gradient_averages.at(i) =
+                past_squared_gradient_decay_rate * past_squared_gradient_averages.at(i) +
+                (1 - past_squared_gradient_decay_rate) * squared_gradient.at(i);
+        }
+
+        // Create the bias corrected gradient and gradient square averages
+        ParamArray bias_corrected_past_gradient_averages         = {0};
+        ParamArray bias_corrected_past_squared_gradient_averages = {0};
+        for (unsigned int i = 0; i < NUM_PARAMS; i++)
+        {
+            bias_corrected_past_gradient_averages.at(i) =
+                past_gradient_averages.at(i) /
+                (1 - std::pow(past_gradient_decay_rate, 2));
+            bias_corrected_past_squared_gradient_averages.at(i) =
+                past_squared_gradient_averages.at(i) /
+                (1 - std::pow(past_squared_gradient_decay_rate, 2));
+        }
+
```

Comment:
Although this step is fairly obvious, a quick comment would be good to be consistent with the previous sections of this function.

Reply:
Sure, added